### PR TITLE
Exploratory field method refactor

### PR DESCRIPTION
**Pull request by ehanson8**
### Purpose and background context
This is intended to start the discussion over how to approach the refactoring of `transmogrifier` to use field methods. Please be thorough and ruthless so we have a good template before kicking off the refactor.

### Includes new or updated dependencies?
NO

### Changes expectations for external applications?
NO

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/TIMX-273

### Developer
- [x] All new ENV is documented in README
- [x] All new ENV has been added to staging and production environments
- [x] All related Jira tickets are linked in commit message(s)
- [x] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed and verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -65,9 +77,50 @@ def datacite_record_all_fields():
...
+
+@pytest.fixture
+def dspace_dim_record_attribute_and_subfield_variations():
+    source_records = DspaceDim.parse_source_file(
+        "tests/fixtures/dspace/dspace_dim_record_attribute_and_subfield_variations.xml"
+    )
+    return next(source_records)
+
+
+@pytest.fixture
```
Comment (ehanson8) [inline tests/conftest.py:99]: Unlike the `_attribute_and_subfield_variations` fixtures (which I didn't touch but have concerns about how useful they are), `_errors` fixtures are intended to hold all edge cases that would trigger logging or alternate behavior. Keeping them all in one fixture that is used by multiple edge case tests (`test_get_dates_invalid_date_range_skipped`) should keep the testing suite cleaner. Open to other names for this fixture

File: tests/conftest.py [hunk 2]
```diff
@@ -65,9 +77,50 @@ def datacite_record_all_fields():
...
+    source_records = DspaceDim.parse_source_file(
+        "tests/fixtures/dspace/dspace_dim_record_errors.xml"
+    )
+    return next(source_records)
+
+
+@pytest.fixture
+def dspace_dim_record_optional_fields_blank():
+    source_records = DspaceDim.parse_source_file(
+        "tests/fixtures/dspace/dspace_dim_record_optional_fields_blank.xml"
```
Comment (ehanson8) [inline tests/conftest.py:110]: We should likely rename this fixture and `_optional_fields_missing` during the refactor

File: tests/sources/xml/test_dspace_dim.py [hunk 3]
```diff
@@ -216,3 +216,48 @@ def test_dspace_dim_transform_with_optional_fields_missing_transforms_correctly(
...
+
+
+def test_get_dates_transforms_correctly_if_fields_missing(
+    dspace_dim_record_optional_fields_missing,
+):
+    assert DspaceDim.get_dates(dspace_dim_record_optional_fields_missing, "abc123") == []
+
+
+def test_get_dates_invalid_date_range_skipped(dspace_dim_record_errors):
+    assert DspaceDim.get_dates(dspace_dim_record_errors, "abc123") == []
```
Comment (ehanson8) [inline tests/sources/xml/test_dspace_dim.py:263]: In addition to `success`, `blank`, and `missing`; tests should be included for any edge cases handled differently by the code. This would vary depending on both the source and field

File: transmogrifier/sources/xml/dspace_dim.py [hunk 4]
```diff
@@ -259,6 +221,73 @@ def get_optional_fields(self, xml: Tag) -> dict | None:
...
+            if "/" in coverage_value:
+                date_object = cls._parse_date_range(coverage_value, source_record_id)
+            else:
+                date_object = timdex.Date(note=coverage_value, kind="coverage")
+            if date_object:
+                coverage_dates.append(date_object)
+        return coverage_dates
+
+    @classmethod
+    def _parse_date_range(
```
Comment (ehanson8) [inline transmogrifier/sources/xml/dspace_dim.py:270]: In the first pass, I propose minimizing the amount of refactoring to creating the field method from the appropriate `get_optional_fields` code block with the exception of overly large code blocks such as this. I don't know how many there will be but I expect `dates` will be a common culprit

File: transmogrifier/sources/xml/dspace_dim.py [hunk 5]
```diff
@@ -259,6 +221,73 @@ def get_optional_fields(self, xml: Tag) -> dict | None:
 
         return fields
 
+    @classmethod
+    def get_contents(cls, xml: Tag) -> list[str]:
+        return [
+            str(contents.string)
+            for contents in xml.find_all(
```
Comment (ehanson8) [inline transmogrifier/sources/xml/dspace_dim.py:228]: Per https://mitlibraries.atlassian.net/browse/TIMX-276, we should soon evaluate whether we want to use `lxml` or `BeautifulSoup` to parse XML

File: tests/sources/xml/test_dspace_dim.py [hunk 6]
```diff
@@ -216,3 +216,48 @@ def test_dspace_dim_transform_with_optional_fields_missing_transforms_correctly(
...
+
+
+def test_get_contents_transforms_correctly_if_fields_missing(
+    dspace_dim_record_optional_fields_missing,
+):
+    assert DspaceDim.get_contents(dspace_dim_record_optional_fields_missing) == []
+
+
+def test_get_dates_success(dspace_dim_record_all_fields):
+    assert DspaceDim.get_dates(dspace_dim_record_all_fields, "abc123") == [
```
Comment (ehanson8) [inline tests/sources/xml/test_dspace_dim.py:238]: For field method `_success` tests, we should extract the values from the `_all_fields_transforms_correctly` test

**Review by jonavellecuerdo — COMMENTED** [review]
@ehanson8 I think this PR does a great job at explaining **the main goal of the refactor** by providing a few informative examples without making sweeping changes all at once! I also appreciated the comments you added to guide our reviews. 

I proposed a few updates! Looking forward to our team discussion on this first pass! ✨

File: transmogrifier/sources/xml/dspace_dim.py [hunk 7]
```diff
@@ -259,6 +221,73 @@ def get_optional_fields(self, xml: Tag) -> dict | None:
...
+
+    @classmethod
+    def _parse_date_range(
+        cls, coverage_value: Tag, source_record_id: str
+    ) -> timdex.Date | None:
+        """Parse date range value and return a Date object if it is validated."""
+        split = coverage_value.index("/")
+        gte_date = coverage_value[:split]
+        lte_date = coverage_value[split + 1 :]
+        if validate_date_range(
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply transmogrifier/sources/xml/dspace_dim.py:277]: A good point and worthy of discussion, but this does feel like a slightly different type of validation than what we're doing init'ing a `TimdexRecord` which is more checking types rather than the value's content. I'll add it to the agenda!

File: transmogrifier/sources/xml/dspace_dim.py [hunk 8]
```diff
@@ -259,6 +221,73 @@ def get_optional_fields(self, xml: Tag) -> dict | None:
 
         return fields
 
+    @classmethod
+    def get_contents(cls, xml: Tag) -> list[str]:
```
[the remark this answers is not in the document]
Comment (ghukill) [reply transmogrifier/sources/xml/dspace_dim.py:225]: I think I like `source_record` as well.  That certainly handles either a JSONLines parsed `dict` or an XML parsed BS4 `Tag` or lxml `Element` (XML root).

**Review by ghukill — COMMENTED** [review]
Thanks for this first pass at some "field methods" to provide a backdrop for discussions.

My comments contain the bulk of my thoughts so far, but to summarize here and add a couple others:

- thinking we might want to consider small, stubbed `source_record` instances that are defined in each test vs. complicated fixtures
- think the decision of BS4 vs lxml will have some bearing, but not a lot, on what field methods look like (e.g. if we're using XPath, to consider some helper methods)
- very curious to talk about orchestration of these field methods, as I'm confident that will have some impact on their ergonomics
- if a field method requires sub-methods as helpers, think about naming them in relation to the part of the data structure they are accessing if appropriate... or a variety of the TIMDEX field they further breaking down if more applicable.  I think this will be hard to get "right", so probably just emerge over time.

Overall though, I think the `get_contents()` and `get_dates()` are looking like what I had hoped and expected!  Concise, clear methods that pull a value for a particular field.

Looking forward towards more discussion, I think if we _did_ want helper methods on XML or JSON objects for things like Xpath, then it's possible the `source_record` we pass to the field methods might have more capabilities, and therefore also potentially change those signatures.  Specifically, I've always thought it's a little awkward to pass the `source_record_id` into the methods.  If we pass an actual `SourceRecord` object, per say, we could have access to a `source_record.identifier` for example.  Just food for thought in the discussion.

**Comment by ehanson8** [comment]
> Specifically, I've always thought it's a little awkward to pass the `source_record_id` into the methods. If we pass an actual `SourceRecord` object, per say, we could have access to a `source_record.identifier` for example. Just food for thought in the discussion.

Agreed, I hated adding that param in this commit because `source_record.identifier` should be a thing!

File: transmogrifier/sources/xml/dspace_dim.py [hunk 9]
```diff
@@ -259,6 +221,73 @@ def get_optional_fields(self, xml: Tag) -> dict | None:
...
+        for date_element in xml.find_all("dim:field", element="date", string=True):
+            date_value = str(date_element.string.strip())
+            if validate_date(date_value, source_record_id):
+                if date_element.get("qualifier") == "issued":
+                    date_object = timdex.Date(value=date_value, kind="Publication date")
+                else:
+                    date_object = timdex.Date(
+                        value=date_value, kind=date_element.get("qualifier") or None
+                    )
+                dates.append(date_object)
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply transmogrifier/sources/xml/dspace_dim.py:248]: > For example:
> 
> ```python
> def get_dates():
>   dates = []
>   dates.extend(self._get_dates_from_date_elements())
>   dates.extend(self._get_dates_from_coverage_elements())
> ```

I like the logic of that pattern, let's discuss at meeting

File: tests/fixtures/dspace/dspace_dim_record_errors.xml [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+        <metadata>
+            <dim:dim xmlns:dim="http://www.dspace.org/xmlns/dspace/dim"
+                xmlns:doc="http://www.lyncode.com/xoai"
+                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                xsi:schemaLocation="http://www.dspace.org/xmlns/dspace/dim http://www.dspace.org/schema/dim.xsd">
+                <dim:field element="coverage" qualifier="temporal">2020-01-02/2019-01-01</dim:field>
+            </dim:dim>
+        </metadata>
+    </record>
+</records>
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply tests/fixtures/dspace/dspace_dim_record_errors.xml:12]: Good point, slightly worried about messy tests with the XML stubs but let's discuss

File: tests/sources/xml/test_dspace_dim.py [hunk 11]
```diff
@@ -216,3 +216,48 @@ def test_dspace_dim_transform_with_optional_fields_missing_transforms_correctly(
         format="electronic resource",
         content_type=["Not specified"],
     )
+
+
+def test_get_contents_success(dspace_dim_record_all_fields):
+    assert DspaceDim.get_contents(dspace_dim_record_all_fields) == ["Chapter 1"]
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply tests/sources/xml/test_dspace_dim.py:222]: This looks like a good model and the logic is much cleaner than my proposed `_errors` fixture

[the remark this answers is not in the document]
Comment (ghukill) [reply tests/sources/xml/test_dspace_dim.py:222]: Be interested to really kick around in discussion though.  Because, thankfully, I don't think we have many (any?) transforms that look at element `<foo>` _and_ element `<bar>` to determine a value for the final TIMDEX record, I think these XML blocks we insert will be an element or two.  

If we _were_ in the business of taking an element like `<foo>` and then comparing against other aspects of the record that are kind of tangential to that element, but help inform how to interpret, than this pattern could get cumbersome.  But, maybe that'd be a good instance for a pytest fixture.

[the remark this answers is not in the document]
Comment (ehanson8) [reply tests/sources/xml/test_dspace_dim.py:222]: True, I believe `marc` has some examples involving 2 fields but not many

File: tests/sources/xml/test_dspace_dim.py [hunk 12]
```diff
@@ -234,8 +260,21 @@ def test_get_contents_transforms_correctly_if_fields_missing(
     assert DspaceDim.get_contents(dspace_dim_record_optional_fields_missing) == []
 
 
-def test_get_dates_success(dspace_dim_record_all_fields):
-    assert DspaceDim.get_dates(dspace_dim_record_all_fields, "abc123") == [
+def test_get_dates_success():
+    source_record = create_dspace_dim_source_record_stub(
+        '<dim:field mdschema="dc" element="coverage" qualifier="temporal">'
```
[the remark this answers is not in the document]
Comment (ghukill) [reply tests/sources/xml/test_dspace_dim.py:265]: Agreed, this is probably the trickiest part to get ergonomically right.

Here's an option:
```python
# at top of file, blanket skip long lines in a test file, which I'd be okay with
# ruff: noqa: E501

# then in test, can prevent auto-formatting for a block of code with #fmt: off / #fmt: on
source_record = create_dspace_dim_source_record_stub(
    # fmt: off
    """
    <dim:field mdschema="dc" element="coverage" qualifier="temporal">1201-01-01 - 1965-12-21</dim:field>
    <dim:field mdschema="dc" element="coverage" qualifier="temporal">1201-01-01/1965-12-21</dim:field>
    <dim:field mdschema="dc" element="date" qualifier="accessioned">2009-01-08T16:24:37Z</dim:field>
    <dim:field mdschema="dc" element="date" qualifier="available">2009-01-08T16:24:37Z</dim:field>
    <dim:field mdschema="dc" element="date" qualifier="issued">2002-11</dim:field>
    <dim:field mdschema="dc" element="identifier" qualifier="uri">https://hdl.handle.net/1912/2641</dim:field>
    """
    # fmt: on
)
```
